src/enrichment_tables/pgtable.rs

```diff
@@ -0,0 +1,521 @@
+//! Handles enrichment tables for `type = pgtable`.
+use std::{collections::HashMap, hash::Hasher};
+use std::{path::PathBuf};
+use bytes::Bytes;
+use vector_lib::enrichment::{Case, Condition, IndexHandle, Table};
+use itertools::intersperse;
+
+// use bytes::Bytes;
+// use tracing::trace;
+use vector_lib::configurable::configurable_component;
+// use vector_lib::{conversion::Conversion, TimeZone};
+// use vrl::value::{ObjectMap, Value};
+use chrono::{DateTime, Utc};
+use crate::config::EnrichmentTableConfig;
+use snafu::{Snafu};
+use postgres_openssl::MakeTlsConnector;
+use tokio_postgres::{Config, Error as PgError, NoTls, Row, Statement, Client};
+use tokio_postgres::types::Type;
+use openssl::{
+    error::ErrorStack,
+    ssl::{SslConnector, SslMethod},
+};
+use vrl::value::{ObjectMap, Value};
+use futures::executor::block_on;
+
+#[derive(Debug, Snafu)]
+enum ConnectError {
+    #[snafu(display("failed to create tls connector: {}", source))]
+    TlsFailed { source: ErrorStack },
+    #[snafu(display("failed to connect ({}): {}", endpoint, source))]
+    ConnectionFailed { source: PgError, endpoint: String },
+}
+
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("invalid endpoint: {}", source))]
+    InvalidEndpoint { source: PgError },
+    // #[snafu(display("host missing"))]
+    // HostMissing,
+    // #[snafu(display("multiple hosts not supported: {:?}", hosts))]
+    // MultipleHostsNotSupported { hosts: Vec<Host> },
+}
+
+/// Configuration of TLS when connecting to PostgreSQL.
+#[configurable_component]
+#[derive(Clone, Debug)]
+#[serde(deny_unknown_fields)]
+struct PgtableTlsConfig {
+    /// Absolute path to an additional CA certificate file.
+    ///
+    /// The certificate must be in the DER or PEM (X.509) format.
+    #[configurable(metadata(docs::examples = "certs/ca.pem"))]
+    ca_file: PathBuf,
+}
+
+/// Postgres-specific settings
+#[configurable_component(enrichment_table("pgtable"))]
+#[derive(Clone, Debug, Default)]
+pub struct PgtableConfig {
+    #[configurable(derived)]
+    /// The postgres connection string
+    endpoint: String,
+    /// Table to query
+    table: String,
+    /// Columns to project from table
+    columns: Vec<String>,
+    /// Reload the table when this query returns a newer timestamp
+    /// than the last one that was seen. The result set must contain
+    /// zero rows or one row, and it must contain a single timestamptz
+    /// column. The name of the column does not matter.
+    ///
+    /// Example: SELECT max(w.updated) FROM widgets w;
+    reload_when: Option<String>,
```

Comment:
I was thinking that maybe it would be useful to use a `reload_interval` so we could just execute the query and update the table every 10minutes or so.

Some thoughts @StephenWakely ?

Reply:
@jorgehermo9 @andrewthad It will be excellent!

Reply:
From a UX perspective I can definitely see the usefulness.

The reloading occurs at a topology level [here](https://github.com/vectordotdev/vector/blob/master/src/topology/builder.rs#L162). To ensure that access to the enrichment tables can be done in a thread safe manner, without using any locks - the data is stored in an [ArcSwap](https://docs.rs/arc-swap/latest/arc_swap/). If `needs_reload` here returns true then an entirely new set of enrichment table data is loaded in. The existing enrichment table is still around and can continue to serve requests while the new table is being loaded. Once loaded an atomic swap occurs to make the new table the current one and the old table is dropped.

So, all the logic to enable a reload interval would need to be implemented in the topology rather than here.

It would be out of scope for this PR, but definitely worth raising an issue to be looked at in future.